X-HTTP-Method-Override listener

Wilt opened the pull request:
X-HTTP-Method-Override listener feature.
For this issue: https://github.com/zfcampus/zf-apigility/issues/175#issuecomment-241523026

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,85 @@
+
+        $result = $listener($event);
+        $this->assertEquals($method, $request->getMethod());
+
```
Please remove empty line above.

webimpress commented on src/HttpMethodOverrideListener.php:
```diff
@@ -0,0 +1,58 @@
+
+        $method = $header->getFieldValue();
+
+        if (!in_array($method, $this->methods)) {
```
Please add one space after `!` op.

webimpress commented on src/HttpMethodOverrideListener.php:
```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Please update year - `2016` instead of `2014` here.

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,85 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Please update year - `2016` instead of `2014`.

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,85 @@
+        $event->setRequest($request);
+
+        $result = $listener($event);
+        $this->assertInstanceOf('ZF\ApiProblem\ApiProblemResponse', $result);
```
Please use `::class` notation

webimpress commented on src/HttpMethodOverrideListener.php:
```diff
@@ -0,0 +1,58 @@
+        if (!in_array($method, $this->methods)) {
+            return new ApiProblemResponse(new ApiProblem(
+                400,
+                'unrecognized method in X-HTTP-Method-Ovverride header'
```
Please change to uppercase `U` in `unrecognized` word

Wilt commented:
PR for this issue: https://github.com/zfcampus/zf-apigility/issues/175#issuecomment-241523026

Didn't yet implement a solution for connecting the listener. Maybe an optional config param for enabling this listener. I think enabling this listener should be optional, feels not good to connect such a listener by default to the application.

Maybe someone from zf-campus can give some guidance on how they would like to connect it...?

webimpress commented on src/HttpMethodOverrideListener.php:
```diff
@@ -0,0 +1,58 @@
+        if (! in_array($method, $this->methods)) {
+            return new ApiProblemResponse(new ApiProblem(
+                400,
+                'unrecognized method in X-HTTP-Method-Ovverride header'
```
`U` in `unrecognized` please

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,85 @@
+
+    /**
+     *
+     */
```
What does it for?

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,85 @@
+        $this->assertInstanceOf(ApiProblemResponse::class, $result);
+        $problem = $result->getApiProblem();
+        $this->assertEquals(400, $problem->status);
+        $this->assertContains('unrecognized method in X-HTTP-Method-Ovverride header', $problem->detail);
```
`U` in `unrecognized` please

Wilt commented:
Updated the PR:

- Made HttpMethodOverrideListener extend AbstractListenerAggregate
- Attached listener depending on `x_http_method_override_enabled`
- Added $xHttpMethodOverrideEnabled to ContentNegotiationOptions (defaults to false);

snapshotpl commented:
@Wilt there is already existing module https://github.com/rstgroup/http-method-override

Wilt commented:
@snapshotpl OK, thanks. 
If there is no interest to implement this in Apigility/zf-content-negotiation, then someone can just close and I will delete my PR.

webimpress commented on src/Factory/HttpMethodOverrideListenerFactory.php:
```diff
@@ -0,0 +1,27 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014-2016 Zend Technologies USA Inc. (http://www.zend.com)
```
Please use only `2016` year in new files.

webimpress commented on test/Factory/HttpMethodOverrideListenerFactoryTest.php:
```diff
@@ -0,0 +1,31 @@
+
+        $factory = new HttpMethodOverrideListenerFactory();
+
+        $service = $factory($serviceManager, 'HttpMethodOverrideListener');
```
Please use here as second parameter `HttpMethodOverrideListener::class` (I know, it is not used)

webimpress commented on test/Factory/HttpMethodOverrideListenerFactoryTest.php:
```diff
@@ -0,0 +1,31 @@
+        $serviceManager->setService(
+            ContentNegotiationOptions::class,
+            new ContentNegotiationOptions()
+        );
```
You can use `->prophesize(...)` to create service manager mock (please check how it is done in other tests)

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,127 @@
+            HttpRequest::METHOD_PATCH,
+        ],
+        HttpRequest::METHOD_POST => [
+        ]
```
Trailing comma please.

webimpress commented on src/HttpMethodOverrideListener.php:
```diff
@@ -0,0 +1,85 @@
+     * Priority is set very high (should be executed before all other listeners that rely on the request method value).
+     * TODO: Check priority value, maybe value should be even higher??
+     *
+     * @param  EventManagerInterface $events
```
one space after `@param` please

webimpress commented on test/Factory/HttpMethodOverrideListenerFactoryTest.php:
```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
`2016` year please

webimpress commented:
@Wilt I like it, I think will be nice to have this functionality out-of-the-box. 👍

Wilt commented:
@webimpress Thanks for the feedback. 

I changed to prophecy. I don't like the fact that the `get` method and `getHttpOverrideMethods` in my test aren't recognized. Is there a workaround?

webimpress commented on test/HttpMethodOverrideListenerTest.php:
```diff
@@ -0,0 +1,127 @@
+            HttpRequest::METHOD_PATCH
+        ],
+        HttpRequest::METHOD_POST => [
+        ]
```
please add trailing comma

webimpress commented:
@Wilt do you mean your IDE doesn't recognize these methods?
You can always add comment `/** @var YourClass $variable */` before variable, and if you are using PHPStorm it will be recognized properly.
But then it's not 100% true, because prophecy object is `ProphecyInterface`, so maybe better if your comment will be:
`/** @var YourClass|ProphecyInterface $variable */`

jdukleth commented:
Looks great. I think this should be baked into Apigility.

weierophinney commented on src/HttpMethodOverrideListener.php:
```diff
@@ -0,0 +1,85 @@
+
+    /**
+     * Priority is set very high (should be executed before all other listeners that rely on the request method value).
+     * TODO: Check priority value, maybe value should be even higher??
```
This is an appropriate value. The OPTIONS handling (i.e., HTTP method negotiation) happens at -100. Currently, we already have versioning (for Accept headers) happening at -40 (route-based version checking happens at -41). The main thing is that it happens before the OPTIONS handling, so this works fine.

weierophinney commented on src/Module.php:
```diff
@@ -48,6 +48,11 @@ public function onBootstrap(MvcEvent $e)
 
+        $contentNegotiationOptions = $services->get(ContentNegotiationOptions::class);
+        if ($contentNegotiationOptions->getXHttpMethodOverrideEnabled()) {
+            $services->get(HttpMethodOverrideListener::class)->attach($eventManager);
```
I very much like this conditional attachment.

weierophinney commented:
Thanks, @Wilt\!

Wilt commented:
You are very welcome :smiley:
